test/HttpTest.cpp

```diff
@@ -381,3 +381,125 @@ TEST(HttpServer, RequestBodySizeLimit) {
   expectRequestSucceeds(10_kB);
   expectRequestSucceeds(5_MB);
 }
+
+// Test HTTP redirect handling in `sendHttpOrHttpsRequest`.
+TEST(HttpClient, Redirects) {
+  using ::testing::AllOf;
+  ad_utility::SharedCancellationHandle handle =
+      std::make_shared<ad_utility::CancellationHandle<>>();
+
+  // Helper lambda to create a server that redirects to a given location.
+  auto makeRedirectServer = [](status redirectStatus, std::string location) {
+    return TestHttpServer([redirectStatus, location = std::move(location)](
+                              [[maybe_unused]] auto request,
+                              auto&& send) -> boost::asio::awaitable<void> {
+      auto response = []() -> cppcoro::generator<std::string> {
+        co_yield "";
+      }();
+      http::response<http::string_body> resp;
+      resp.result(redirectStatus);
+      resp.set(http::field::content_type, "text/plain");
+      if (!location.empty()) {
+        resp.set(http::field::location, location);
+      }
+      resp.body() = "";
+      resp.prepare_payload();
+      co_await send(std::move(resp));
+    });
+  };
+
+  // Helper lambda to create a server that returns OK.
+  auto makeOkServer = []() {
+    return TestHttpServer(
+        [](auto request, auto&& send) -> boost::asio::awaitable<void> {
+          auto response = []() -> cppcoro::generator<std::string> {
+            co_yield "Success";
+          }();
+          co_return co_await send(createOkResponse(
+              std::move(response), request, ad_utility::MediaType::textPlain));
+        });
+  };
+
+  // Test that all four redirect types (301, 302, 307, 308) work.
+  for (auto redirectStatus :
+       {status::moved_permanently, status::found, status::temporary_redirect,
+        status::permanent_redirect}) {
+    // Create final server first.
+    auto okServer = makeOkServer();
+    okServer.runInOwnThread();
+    std::string finalUrl =
+        absl::StrCat("http://localhost:", okServer.getPort());
+
+    // Create `redirectServer` that points to `okServer`.
```

Comment:
Instead of creating multiple servers that all run in their own threads. Couldn't you just create a single server, that returns a status code based on GET parameters? Like `http://localhost/301?redirect=/200` conceptually? This way you can easily chain them by extending the parameters without having to write much more code.

Reply:
It's not important from a functional perspective, but it avoids the overhead of starting and stopping servers which AFAIK takes rather long for a unit test.
Also I believe it would reduce the amount of code in the test.